### PR TITLE
cmake: rgw: do not link against boost in a wholesale

### PR DESCRIPTION
**Pull request by tchaikov**
Signed-off-by: Kefu Chai <kchai@redhat.com>

### CODE REVIEW

**Comment by smithfarm** [comment]
I guess it's better to keep the PRs separate :-/

**Comment by tchaikov** [comment]
@smithfarm yeah, either way =)

**Comment by smithfarm** [comment]
Inclusion of this patch in #15225 appeared to cause build to fail, that's why I dropped it.

**Comment by tchaikov** [comment]
okay, we can get #15225 in first. then let me worry about this one.

**Comment by wjwithagen** [comment]
@tchaikov 
Not quite sure on how this is going to fxi my `-lpython` problem completely.
I see:
```
./CMakeLists.txt:638:# target_link_libraries(common json_spirit common_utf8 erasure_code rt uuid ${LIB_RESOLV} ${CRYPTO_LIBS} ${Boost_LIBRARIES} ${BLKID_LIBRARIES} ${EXECINFO_LIBRARIES} ${BLKIN_LIBRARIES})
./rgw/CMakeLists.txt:147:  ${CURL_LIBRARIES} ${Boost_LIBRARIES}
```

This way this list still suffers from my problem: `libboost_python` is on the list and not again linked to `-lpython` generating linking errors.

Where as I'm trying to prevent all this by removing `libboost_python` from the list.

**Comment by tchaikov** [comment]
@wjwithagen 
```
./CMakeLists.txt:638:# target_link_libraries(common json_spirit common_utf8 erasure_code rt uuid ${LIB_RESOLV} ${CRYPTO_LIBS} ${Boost_LIBRARIES} ${BLKID_LIBRARIES} ${EXECINFO_LIBRARIES} ${BLKIN_LIBRARIES})
./rgw/CMakeLists.txt:147:  ${CURL_LIBRARIES} ${Boost_LIBRARIES}
```

is addressed by 
https://github.com/ceph/ceph/pull/15225/files#diff-95e351a3805a1dafa85bf20b81d086e6L640

**Comment by smithfarm** [comment]
@tchaikov Please rebase

**Comment by smithfarm** [comment]
@tchaikov Nevermind - I'll cherry-pick for the OBS test, but for Shaman test you should rebase.

**Comment by tchaikov** [comment]
@smithfarm rebased and pushed to ci as `wip-15347-kefu`.

**Comment by tchaikov** [comment]
@smithfarm the branch builds fine on shaman, what about OBS?

**Comment by smithfarm** [comment]
@tchaikov Sorry for the latency. On it now.

**Comment by smithfarm** [comment]
It takes quite some time for OBS to build for all the target platforms/architectures. Expect to hear the result no earlier than tomorrow.

**Comment by tchaikov** [comment]
@smithfarm thanks for your time!

**Comment by smithfarm** [comment]
@tchaikov The SUSE build fails with the following error:

```
[ 2264s] [100%] Linking CXX executable ../../bin/radosgw
[ 2270s] ../../boost/lib/libboost_coroutine.a(coroutine_context.o): In function `boost::coroutines::detail::coroutine_context::coroutine_context(void (*)(boost::context::detail::transfer_t), boost::coroutines::detail::preallocated const&)':
[ 2270s] coroutine_context.cpp:(.text+0x59): undefined reference to `make_fcontext'
[ 2270s] ../../boost/lib/libboost_coroutine.a(coroutine_context.o): In function `boost::coroutines::detail::coroutine_context::jump(boost::coroutines::detail::coroutine_context&, void*)':
[ 2270s] coroutine_context.cpp:(.text+0xe8): undefined reference to `jump_fcontext'
[ 2270s] collect2: error: ld returned 1 exit status
[ 2270s] src/rgw/CMakeFiles/radosgw.dir/build.make:159: recipe for target 'bin/radosgw' failed
[ 2270s] make[2]: *** [bin/radosgw] Error 1
[ 2270s] CMakeFiles/Makefile2:7841: recipe for target 'src/rgw/CMakeFiles/radosgw.dir/all' failed
[ 2270s] make[1]: *** [src/rgw/CMakeFiles/radosgw.dir/all] Error 2
[ 2270s] make[1]: *** Waiting for unfinished jobs.... 
```

I notice that the SUSE verion of `FindBoost.cmake` does not appear to set `Boost_<C>` at all. Where are these variables `Boost_COROUTINE` and `Boost_CONTEXT` supposed to be getting set?

Could we use `Boost_<C>_LIBRARY` instead, and possibly guard it with `Boost_<C>_FOUND`?

**Comment by tchaikov** [comment]
i don't think we need to protect them with `Boost_<C>_FOUND`, as `find_package(Boost 1.61 COMPONENTS ${BOOST_COMPONENTS} **REQUIRED**)` has taken care of it.

**Comment by smithfarm** [comment]
@tchaikov Getting the same failure now with ceph_test_cephd_api_misc -> https://github.com/tchaikov/ceph/pull/2

**Comment by smithfarm** [comment]
We verified that linking ceph_test_cephd_api_misc with `${Boost_LIBRARIES}` fixes it, now testing this version.

**Comment by tchaikov** [comment]
@smithfarm i am fixing it in a slightly different way. does it make sense to you? and i also added your "Signed-off-by" line. is it fine with you?

**Comment by smithfarm** [comment]
@tchaikov You did what I was trying to do - I trust it's OK. We're testing it now.

**Comment by smithfarm** [comment]
```
100%] Linking CXX executable ../../bin/radosgw-admin
../../boost/lib/libboost_coroutine.a(coroutine_context.o): In function `boost::coroutines::detail::coroutine_context::coroutine_context(void (*)(boost::context::detail::transfer_t), boost::coroutines::detail::preallocated const&)':
coroutine_context.cpp:(.text+0x59): undefined reference to `make_fcontext'
../../boost/lib/libboost_coroutine.a(coroutine_context.o): In function `boost::coroutines::detail::coroutine_context::jump(boost::coroutines::detail::coroutine_context&, void*)':
coroutine_context.cpp:(.text+0xe5): undefined reference to `jump_fcontext'
collect2: error: ld returned 1 exit status
src/rgw/CMakeFiles/radosgw.dir/build.make:161: recipe for target 'bin/radosgw' failed
make[2]: *** [bin/radosgw] Error 1
CMakeFiles/Makefile2:21995: recipe for target 'src/rgw/CMakeFiles/radosgw.dir/all' failed
make[1]: *** [src/rgw/CMakeFiles/radosgw.dir/all] Error 2
```

**Comment by smithfarm** [comment]
So `rgw_admin.cc` does use the `RGWCoroutinesManager` class -> https://github.com/ceph/ceph/blob/master/src/rgw/rgw_admin.cc#L6240

**Comment by cbodley** [comment]
@smithfarm `RGWCoroutinesManager` and friends in `rgw_coroutine.h` are based on `boost::asio::coroutine` from `boost/asio.hpp`, which is completely separate from the boost coroutines library

the only thing that should depend on the coroutines/context libs is the `boost/asio/spawn.hpp` header used by `rgw_asio_frontend.cc`

**Comment by smithfarm** [comment]
@cbodley Thanks for the hints!

**Comment by tchaikov** [comment]
@smithfarm thanks for your efforts! squashed and repushed.

File: src/rgw/CMakeLists.txt [hunk 1]
```diff
@@ -260,4 +265,9 @@ if(WITH_EMBEDDED)
   if(WITH_RADOSGW_FCGI_FRONTEND) 
     target_link_libraries(cephd_rgw ${FCGI_LIBRARY})
   endif()
+  if(WITH_RADOSGW_BEAST_FRONTEND)
+    target_link_libraries(cephd_rgw LINK_PRIVATE
+      ${Boost_COROUTINE_LIBRARY}
+      ${Boost_CONTEXT_LIBRARY})
+  endif()
```
Comment (smithfarm) [inline src/rgw/CMakeLists.txt]: I suspect this stanza is unnecessary. Running a `-DWITH_EMBEDDED=ON` test build without the stanza.

Comment (tchaikov) [inline src/rgw/CMakeLists.txt]: you mean the whole 
```cmake
    target_link_libraries(cephd_rgw LINK_PRIVATE
      ${Boost_COROUTINE_LIBRARY}
      ${Boost_CONTEXT_LIBRARY})
```
block?

Comment (smithfarm) [inline src/rgw/CMakeLists.txt]: Yup. I built with `./do_cmake.sh -DCMAKE_INSTALL_PREFIX=/usr -DCMAKE_INSTALL_LIBDIR=/usr/lib64 -DCMAKE_INSTALL_LIBEXECDIR=/usr/lib -DCMAKE_INSTALL_LOCALSTATEDIR=/var -DCMAKE_INSTALL_SYSCONFDIR=/etc -DCMAKE_INSTALL_MANDIR=/usr/share/man -DCMAKE_INSTALL_DOCDIR=/usr/share/doc/packages/ceph -DWITH_EMBEDDED=ON -DWITH_MANPAGE=ON -DWITH_PYTHON3=ON -DWITH_SYSTEMD=ON -DWITH_TESTS=OFF -DWITH_LTTNG=ON -DHAVE_BABELTRACE=ON -DWITH_OCF=ON -DWITH_RADOSGW_BEAST_FRONTEND=ON -DBOOST_J=8` and it succeeds without this block.

Comment (smithfarm) [inline src/rgw/CMakeLists.txt]: Oh, wait - need to include `-DWITH_TESTS=ON` . . . let me do another build.

Comment (tchaikov) [inline src/rgw/CMakeLists.txt]: yes, need to make sure the `ceph_test_cephd_api_misc` links.

Comment (smithfarm) [inline src/rgw/CMakeLists.txt]: @tchaikov: Confirmed, see https://github.com/tchaikov/ceph/pull/4

**Comment by tchaikov** [comment]
pushed to ceph-ci as wip-15347-kefu

**Comment by smithfarm** [comment]
Nit: maybe change title to `cmake: rgw: do not link against boost in a wholesale`

**Comment by tchaikov** [comment]
done!

**Comment by tchaikov** [comment]
```
make: *** [all] Error 2
error: Bad exit status from /var/tmp/rpm-tmp.HV6F3b (%build)


RPM build errors:
    Bad exit status from /var/tmp/rpm-tmp.HV6F3b (%build)
```

https://shaman.ceph.com/builds/ceph/wip-15347-kefu/801b06fa3f35d851de525343effcc2df4101dca9/notcmalloc/45824/

**Comment by smithfarm** [comment]
@tchaikov Yeah, that failure is very strange. See further up:

```
[ 44%] Building CXX object src/mds/CMakeFiles/mds.dir/Migrator.cc.o
{standard input}: Assembler messages:
{standard input}:406703: Warning: partial line at end of file ignored
{standard input}:406649: Error: invalid operands (*UND* and .gcc_except_table sections) for `-'
c++: internal compiler error: Killed (program cc1plus)
Please submit a full bug report,
with preprocessed source if appropriate.
See <http://bugzilla.redhat.com/bugzilla> for instructions.
make[2]: *** [src/mds/CMakeFiles/mds.dir/MDCache.cc.o] Error 4
make[2]: *** Waiting for unfinished jobs....
```

**Comment by tchaikov** [comment]
```
[ 86%] Building CXX object src/tools/cephfs/CMakeFiles/cephfs-journal-tool.dir/JournalFilter.cc.o
{standard input}: Assembler messages:
{standard input}:206621: Warning: end of file not at end of a line; newline inserted
{standard input}:207482: Error: character following name is not '#'
c++: internal compiler error: Killed (program cc1plus)
Please submit a full bug report,
with preprocessed source if appropriate.
See <http://bugzilla.redhat.com/bugzilla> for instructions.
make[2]: *** [src/test/rbd_mirror/CMakeFiles/unittest_rbd_mirror.dir/image_replayer/test_mock_BootstrapRequest.cc.o] Error 4
```

again, mysterious error. i am inclined to believe this is an env issue.

**Comment by tchaikov** [comment]
and this also happens with wip-bluestore branch
```
[ 40%] Built target common-objs
make: *** [all] Error 2
error: Bad exit status from /var/tmp/rpm-tmp.TNtJsM (%build)


RPM build errors:
    Bad exit status from /var/tmp/rpm-tmp.TNtJsM (%build)
+ rm -fr /tmp/install-deps.20824
Build step 'Execute shell' marked build as failure
```

https://shaman.ceph.com/builds/ceph/wip-bluestore/ad9bd66ea3873c91ad12e99b75f3c9408167ce31/default/45828/


@yuriw i vaguely remember that you were mumbling the same issue on IRC last week.

@alfredodeza is it possible it's an issue in the jenkins builder?

**Comment by smithfarm** [comment]
Yup, other branches are suffering from the same problem - see e.g. https://shaman.ceph.com/builds/ceph/wip-bluestore/ad9bd66ea3873c91ad12e99b75f3c9408167ce31/default/45828/

**Comment by smithfarm** [comment]
Looks like it started 14 hours ago (about 6:00 p.m. EST Friday June 2) with this build failure: https://shaman.ceph.com/builds/ceph/wip-sage-testing2/f93ad23a8fec219667a03695136842edb0cceace/default/45729/

**Comment by tchaikov** [comment]
```
[ 14%] Building CXX object src/cls/CMakeFiles/cls_cephfs.dir/cephfs/cls_cephfs.cc.o
{standard input}: Assembler messages:
{standard input}:186778: Warning: end of file not at end of a line; newline inserted
{standard input}: Error: open CFI at the end of file; missing .cfi_endproc directive
c++: internal compiler error: Killed (program cc1plus)
Please submit a full bug report,
with preprocessed source if appropriate.
See <http://bugzilla.redhat.com/bugzilla> for instructions.
make[2]: *** [src/mon/CMakeFiles/osd_mon_objs.dir/Monitor.cc.o] Error 4
make[1]: *** [src/mon/CMakeFiles/osd_mon_objs.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

yeah, it does not look right. and we need to fix it.

**Pull request merged**